analyzer/src/types/class.rs

```diff
@@ -0,0 +1,636 @@
+    pub generic_parameters: Vec<GenericParam>,
+
+    ///The associations between child and parent type parameters
+    /// This vector must match the length of parent's type parameters.
+    /// - `class List[A]: Iterable[A]`   The generic param of child (`List`) is associated with generic param of parent (`Iterable`),
+    ///                                  Thus, we have `List::A => Iterable::A`
+    ///
+    /// - `class WeirdMap[A]: Map[A, A]` Here, we have `WeirdMap::A => Map::K` and `WeirdMap::A => Map::V`
+    ///
+    /// - `class Map[K, V]: Iterable[A]` This defines a map that is iterable only on keys, the first map's generic parameter `K`
+    ///                                  is then bound on lone iterable's generic parameter, and the generic param `V` has no links with the parent,
+    ///                                  Thus we only have `Map::K => Iterable::A` as an association set.
+    ///
+    /// - `class StrMap[V]: Map[Str, V]` Here, the associations are `lang::Str => Map::K` and `StrMap::V => Map::V`
+    pub(in crate::types) super_params_associations: Vec<ParamAssociation>,
+
+    ///This Type's identity.
+    pub(crate) identity: u64,
+
+    ///The type context of this type.
+    /// The context contains generic parameters definitions
+    pub context: Rc<RefCell<TypeContext>>,
+}
+
+impl Debug for TypeClass {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TypeClass")
+            .field("name", &self.name)
+            .field("super_type", &self.super_type)
+            .field("generic_parameters", &self.generic_parameters)
+            .field("super_params_associations", &self.super_params_associations)
+            .field("identity", &self.identity)
+            .field("context", &"<not shown>")
+            .finish()
+    }
+}
+
+impl Display for TypeClass {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name)?;
+        if let Some((first, tail)) = self.generic_parameters.split_first() {
+            write!(f, "[{}", first)?;
+            for gparam in tail {
+                write!(f, ", {}", gparam)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+///Different kind of associations between a children and its parent generic parameters
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParamAssociation {
+    ///The association is a known type class (can be a generic)
+    Defined(Rc<TypeClass>),
+    ///The association is nothing
+    Nothing,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GenericParam {
+    pub name: String,
+    pub bound: DefinedType,
+}
+
+impl Display for GenericParam {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}: {}", self.name, self.bound)
+    }
+}
+
+pub struct ClassTypeDefinition {
+    super_type: Option<Rc<TypeClass>>,
+    pub name: String,
+    generic_parameters: Vec<GenericParam>,
+    associations: HashMap<usize, Type>,
```

Comment:
It seems wrong to have a `HashMap` indexed like a `Vec`.

Reply:
This has been done to force the user to explicitly write what index of parent's generic is associated to enhance readability, I'm not a real fan of using a vec here as this would still valid if the changes the parent's type signature